datasetinsights/evaluation_metrics/average_recall_2d.py

```diff
@@ -0,0 +1,133 @@
+r"""Reference.
+
+http://cocodataset.org/#detection-eval
+https://arxiv.org/pdf/1502.05082.pdf
+https://github.com/rafaelpadilla/Object-Detection-Metrics/issues/22
+"""
+import collections
+
+import numpy as np
+
+from datasetinsights.io.bbox import group_bbox2d_per_label
+
+from .base import EvaluationMetric
+from .records import Records
+
+
+class AverageRecall(EvaluationMetric):
+    """2D Bounding Box Average Recall metrics.
+
+    Args:
+        iou_threshold (float): iou threshold (default: 0.5)
+        max_detections (int): max detections per image (default: 100)
+    """
+
+    TYPE = "metric_per_label"
+
+    def __init__(self, iou_threshold=0.5, max_detections=100):
+        self._label_records = collections.defaultdict(
+            lambda: Records(iou_threshold=self._iou_threshold)
+        )
+        self._gt_bboxes_count = collections.defaultdict(int)
+        self._iou_threshold = iou_threshold
+        self._max_detections = max_detections
+
+    def reset(self):
+        """Reset AR metrics."""
+        self._label_records = collections.defaultdict(
+            lambda: Records(iou_threshold=self._iou_threshold)
+        )
+        self._gt_bboxes_count = collections.defaultdict(int)
+
+    def update(self, mini_batch):
+        """Update records per mini batch.
+
+        Args:
+            mini_batch (list(list)): a list which contains batch_size of
+            gt bboxes and pred bboxes pair in each image.
+            For example, if batch size = 2, mini_batch looks like:
+            [[gt_bboxes1, pred_bboxes1], [gt_bboxes2, pred_bboxes2]]
+            where gt_bboxes1, pred_bboxes1 contain gt bboxes and pred bboxes
+            in one image
+        """
+        for bboxes in mini_batch:
+            gt_bboxes, pred_bboxes = bboxes
+            for gt_bbox in gt_bboxes:
+                self._gt_bboxes_count[gt_bbox.label] += 1
+
+            pred_bboxes = sorted(
+                pred_bboxes, key=lambda bbox: bbox.score, reverse=True
+            )[: self._max_detections]
```

Comment:
@BlairLee This will fail if `self._max_detections < len(pred_bboxes)`. Also, this behavior should be well documented in `AverageRecall` class docstring.

Reply:
Actually, if `self._max_detections < len(pred_bboxes)`, it would keep all the boxes and no error would be raised. Will add docstring.